src/withUASTEditor.js

```diff
@@ -0,0 +1,115 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import transformer, { DEFAULT_EXPAND_LEVEL } from './transformer';
+import PositionIndex from './PositionIndex';
+import {
+  hoverNodeById,
+  highlightNodeById,
+  makePositionIndexHook,
+  getNodePosition
+} from './helpers';
+
+function withUASTEditor(WrappedComponent) {
+  class UASTEditorWrapper extends Component {
+    constructor(props) {
+      super(props);
+
+      this.state = this.stateFromUast(props.uast);
+
+      this.onNodeHover = this.onNodeHover.bind(this);
+      this.onNodeToggle = this.onNodeToggle.bind(this);
+      this.onCursorChanged = this.onCursorChanged.bind(this);
+    }
+
+    componentWillReceiveProps(nextProps) {
+      if (nextProps.uast !== this.props.uast) {
+        this.setState(this.stateFromUast(nextProps.uast));
+      }
+    }
+
+    stateFromUast(json) {
+      this.posIndex = new PositionIndex();
+
+      return {
+        uast: transformer(
+          json,
+          DEFAULT_EXPAND_LEVEL,
+          makePositionIndexHook(this.posIndex)
+        ),
+        // control highlighted node
+        lastHighlighted: null,
+        // position of hovered node
+        pos: null
+      };
+    }
+
+    onNodeHover(id, prevId) {
```

Comment:
In the demo I'm seeing something strange.

The node hover effect is persistent, when the mouse exists the UAST tree you still get the highlighted text in the editor. But then you click in the text, and a different UAST node gets highlighted. I think there should be a way to clear state.pos when the mouse exists the UAST tree.

Reply:
fixed